keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1794 @@
+
+Among these modeled initial values, we would get between 3-7 excess restarts per
+backoff lifetime, mostly within the first three time windows matching today's
+restart behavior.
+
+### Rapid curve methodology
+
+For some users in
+[Kubernetes#57291](https://github.com/kubernetes/kubernetes/issues/57291), any
+delay over 1 minute at any point is just too slow, even if it is legitimately
+crashing. A common refrain is that for independently recoverable errors,
+especially system infrastructure events or recovered external dependencies, or
+for absolutely nonnegotiably critical sidecar pods, users would rather poll more
+often or more intelligently to reduce the amount of time a workload has to wait
+to try again after a failure. In the extreme cases, users want to be able to
+configure (by container, node, or exit code) the backoff to close to 0 seconds.
+This KEP considers it out of scope to implement fully user-customizable
+behavior, and too risky without full and complete benchmarking to node stability
+to allow legitimately crashing workloads to have a backoff of 0, but it is in
+scope for the first alpha to provide users a way to opt workloads in to a even
+faster restart behavior.
+
+The finalization of the initial and max cap can only be done after benchmarking.
+But as a conservative first estimate for alpha in line with maximums discussed
+on [Kubernetes#57291](https://github.com/kubernetes/kubernetes/issues/57291),
+the initial curve is selected at initial=250ms / cap=1 minute, but during
+benchmarking this will be modelled against kubelet capacity, potentially
+targeting something closer to an initial value near 0s, and a cap of 10-30s. To
+further restrict the blast radius of this change before full and complete
+benchmarking is worked up, this is gated by a separate alpha feature gate and is
+opted in to per Pod using a new `restartPolicy: Rapid` value, described below.
+
+
+#### New OneOf for `restartPolicy` -- `Rapid`
+`restartPolicy` is an immutable field in podSpec and containerSpec. If set in
+podSpec, each container in the Pod inherits the Pod's restart policy of either
+`Never` (default), `OnFailure`, or `Always`; for a Job, the only valid options
+are `Never` and `OnFailure`. In containerSpec, it is valid ONLY on init
+containers and ONLY as `Always`, to configure a sidecar container that runs
+continuously alongside the regular containers in the Pod.
+
+This KEP will support a new value for this field, `Rapid`, which on feature flag
+disablement will be interpreted as `Always`. If `restartPolicy: Rapid` is set or
+inherited for a container and the feature flag for this feature is turned on for
+a given cluster, that container will follow the new Rapid backoff curve.
+
+Due to configuring this as another option to this field, this would make Rapid
+backoff possible for restartable init (aka sidecar) containers, Pods,
+Deployments, StatefulSets, ReplicaSets, DaemonSets, but NOT pure init
+containers, Jobs or CronJobs.
```

Comment:
Preserved (as unresolved, since worth discussing) in [ee72b80](https://github.com/kubernetes/enhancements/pull/4604/commits/ee72b80043a1b5b5dc23667a8015d70d494f7e08)